### ISSUE TITLE
View support on distributed queries

### ISSUE THREAD

**Issue by metdos** [citusdata/citus#442]
Currently, views on distributed queries do not work.

We are able to create a view on a distributed query;

``` sql
CREATE VIEW book_reviews AS
  (SELECT * FROM customer_reviews WHERE product_group = 'Book' LIMIT 1000);
```

But if we try to use the view, Citus errors out;

``` sql
SELECT COUNT(1) FROM book_reviews;

ERROR:  cannot plan queries that include both regular and partitioned relations
```

**Comment by metdos** [comment]
One workaround would be to use a UDF or a plpgsql function to wrap the view query.

Option-1:
```sql
CREATE VIEW book_reviews AS 
   (SELECT customer_reviews_product_group_udf()); 
```

where the user-defined function customer_reviews_product_group_udf() runs the
```sql
SELECT * FROM customer_reviews WHERE product_group = 'Book' LIMIT 1000;
```

Option-2:
Use a more generic UDF with the only drawback that the types need to be
specified:
```sql
SELECT * FROM view_wrapper('SELECT * FROM customer_reviews WHERE ... ')
v(a int, b int, ...);
```

Option-3:
Create a plpgsql executor function and use it.

```sql
CREATE OR REPLACE FUNCTION execute(query text)
 RETURNS SETOF record
 LANGUAGE plpgsql
AS $function$
BEGIN
 RETURN QUERY EXECUTE query;
END;
$function$;
```

and then run the query 

```sql
SELECT COUNT(1) FROM 
execute('SELECT * FROM customer_reviews WHERE product_group = ''Book'' LIMIT 1000')
result_set_name(customer_id TEXT,
    review_date DATE,
    review_rating INTEGER,
    review_votes INTEGER,
    review_helpful_votes INTEGER,
    product_id CHAR(10),
    product_title TEXT,
    product_sales_rank BIGINT,
    product_group TEXT,
    product_category TEXT,
    product_subcategory TEXT,
    similar_product_ids CHAR(10)[]);

instead of
SELECT COUNT(1) FROM book_reviews;
```

Note that all of these approaches have the drawback that we need to pull all the data returned by view
 to master and apply aggregates and filters on the master node.

**Comment by metdos** [comment]
Another suggestion from @marcocitus is that "if the view can be partitioned in the same way as the table then you could also create a distributed table in which every shard is a view, e.g.:"
https://gist.github.com/marcocitus/59e46fea102f57ecf27fb12b80294433

Then you can create a view using:
```sql
CREATE TABLE book_reviews SELECT * FROM customer_reviews LIMIT 0;
SELECT master_create_distributed_table('book_reviews', 'customer_id', 'hash');
SELECT create_view_shards('book_reviews', 'customer_reviews', 'SELECT
* FROM %I WHERE product_group = ''Book''');
SELECT count(*) FROM book_reviews;
```
%I is replaced by the name of the customer_reviews shard.

The advantage is that additional filters etc. do get pushed down, but lots of stuff won't work with view tables.

I played with this a bit, and added dbname to make it work on my machine;
https://gist.github.com/metdos/b9b128402ec295b62a788628f9ac8e23

It is a very fast sketch and has some bugs, but I quite liked the idea and wanted to share it.

**Comment by saicitus** [comment]
+2 
Saw this issue arise in 2 SE engagements.

**Comment by marcocitus** [comment]
I wonder if we can detect whether a view can be distributed using the INSERT/SELECT logic @onderkalaci is working on.

**Comment by anarazel** [comment]
On 2016-09-30 08:39:00 -0700, Marco Slot wrote:
> I wonder if we can detect whether a view can be distributed using the INSERT/SELECT logic @onderkalaci is working on.

Not easily I think. The issue is that the query we get in planner hook
have already been rewritten (i.e. views/reules have been expanded via
QueryRewrite). That unfortunately makes it harder, because they don't
necessarily deparse into what we'd expect.

There's a second, pretty easy to resolve, issue: We currently error out
because the view still has an (unreferenced) RTE in the query, and is
considered local.

Other than fixing those two, of which one is a lot harder, it should be
fairly simple (at least for router, but even logical planner shouldn't
be too bad).

**Comment by anarazel** [comment]
I think the way to implement this, for select only, is basically:
- add ruleutils support for adding relation names for views - probably that's just adding an ```rtekind == RTE_SUBQUERY ``` branch in ```get_from_clause_item()``` setting ```printalias``` to true (that's never reached in postgres proper)
- Skip over views in router planner, that can e.g. easily done in UpdateRelationNames()
- Prevent normal real-time/task-tracker planner from looking at views, e.g. by checking for ```relkind == 'v'``` in ```TableEntryList()```.
- Modify ```NeedsDistributedPlanning()``` to not count views

Seems kind of simple, actually.

**Comment by mtuncer** [comment]
some addition to @anarazel's last comment on ```TableEntryList()```
This function blindly returns all RTEs regardless of if they were actually used or not.  A simple select query on a view may end up with 4 RTEs in which only one of them is actually used. Remaining 3 are one subquery RTE and 2 view RTEs. 

Perhaps we should ignore unreferenced RTEs.

**Comment by mtuncer** [comment]
Detecting whether a view is being modified is not trivial.  I am using a not-so-pretty trick.
when you have
```sql
insert into some_view select * from distributed_table
```
Query object contains 3 RTEs
1- view
2- select subquery
3- relation range table entry that view is pointing to

Query->resultRelation points to the relation not to the view. For modify queries, we could check the first RTE relkind to see if it is a view, than bail out  if it is.

**Comment by begriffs** [comment]
Requires documentation in https://github.com/citusdata/citus_docs/issues/227

**Comment by mtuncer** [comment]
Now we have an implementation at hand, it has some drawbacks that we should agree on before merging.

## View Creation
No problems there, user can create a view on distributed table same as on a local table.

## Usage
We have different behavior here.
### Router Plannable queries involving views
If the query involving the view is router plannable (i.e. single worker query on a single node and single tuple of shards) we can execute the view. 
### Non-router plannable 
If a view is simple projection or selection type (```select <some columns> from some_table where <some_filter>```) then we can do simple selects, joins (inner ones, and some outer joins when the view is in the inner part of the join). 

However, when a view has aggregations and group by's, then even simple select queries on the view becomes a subquery, and rejected.

Subquery comes to picture here because Postgresql re-writes the view as a subquery in before the planner, so when we have access to Query object, view is already converted into a subquery.

Reason between different behavior is that simple select views get flattened and subquery is removed from the query by standard planner; whereas if a view has a group by then standard planner does not do much and sends the query as it was.

### Solution Alternatives
I really don't have a solution here. Possible ones are 
- leave it as it is and have views for on router queries. And simple select views for the rest.
- made logical planner to check for views and only let queries with simple views to proceed, reject the other ones with proper error message. Currently (this pr) we give subquery with views are not supported error. Which is kind of misleading.
-  extend subquery support such that it supports **some** views with aggregates/group by 

Decision lies on the use case we are targeting for.  May be we should re-iterate on this.

**Comment by mtuncer** [comment]
After discussion between @anarazel and @ozgune we decided to move forward with this work. View support for real time/task-tracker queries when there is group by in the view would be somewhat limited.

**Comment by lfittl** [comment]
@mtuncer Question, as it stands, two cases I'm curious about:

1) Do we support a view of a UNION query, where each of the individual SELECTs goes to the same tenant? (i.e. all the data is co-located)

2) Do we support a view of a UNION query, where each of the individual SELECTs goes to different tenants? (i.e. the data is only co-located within an individual SELECT, so we are accessing different nodes for each SELECT query)

**Comment by mtuncer** [comment]
@lfittl 
1 - query involving the view is a router query (single tenant + co-located) then yes. 
2 - this involves subquery + union which we do not support at the moment for real-time queries. So the answer is no.


So it all comes down to final query being router plannable if unions are involved.

**Comment by mtuncer** [comment]
needs documentation

**Comment by mtuncer** [comment]
documentation is updated at https://github.com/citusdata/citus_docs/issues/227

**Comment by ozgune** [comment]
@mtuncer -- Did we capture the limitations of our view support in an issue? Is #1071 the right issue to track this?

**Comment by mtuncer** [comment]
@ozgune No we haven't #1071 is just a bug related to subquery processing but can only be visible when a view is  used.

limitations on view support would be the same as limitations of subqueries we are currently working on. We haven't documented them yet.

**Comment by ozgune** [comment]
@mtuncer -- When we say view and subquery support have the same limitations, is that true from the user's perspective?

I'm copy / pasting an example chat session with a user.

---

When you define tables and views:

<pre>
create table t (n int, m int);
select create_distributed_table('t', 'n');
create view v as select * from t;
create view v2 as select n from v group by n;
</pre>

Any select like: select * from v2; results with:
<pre>
ERROR:  cannot perform distributed planning on this query
DETAIL:  Subqueries without aggregates are not supported yet
</pre>

But the weirdest part is that you can run without any problem the query inside the v2 definition, i.e. select n from v group by n;

When that query is wrapped by the view (v2) suddenly it's not queryable.

---

In other words, the confusion here is that Citus supports ``select n from v group by n;``. However, when you wrap this query in a view, it stops working.

If #1071 doesn't cover current limitations, I'll re-open this issue and remove the 6.1 tag. If you'd like, please feel free to capture current limitations in a separate issue (instead of this one).

**Comment by metdos** [comment]
@onderkalaci, @mtuncer  can we close this?

**Comment by onderkalaci** [comment]
>@onderkalaci, @mtuncer can we close this?

I think no. Because we've only increased the view support as a side effect of subquery pushdown changes. There are still many cases that Citus does not support.

The rule of thumb: `If the query that is being used to create the View can be planned by the subquery pushdown planner, then Citus supports that View`

So, for example subquery pushdown planner cannot plan aggregates on non-partition keys. So, such views are not supported. I think until we have a full SQL coverage in subqueries, it is hard to support all kinds of views. (At least if we don't want to do anything special for views, only let Postgres handle everything for us and Citus just plans & executes the subquery that comes from the view)

**Comment by marcocitus** [comment]
All the things mentioned in this thread are supported now. 

The main open issue with views is that you cannot give read access to a role that does not have read access to a table (#1683). This may require explicitly creating views on each shard.
